modules/cart/src/Tests/AddToCartFormTest.php

```diff
@@ -146,128 +147,238 @@ public function testMultipleVariationsAddToCartForm() {
+      '7' => '7in',
+      '8' => '8in',
+      '9' => '9in',
+      '10' => '10in',
+    ]);
+    $color_attributes = $this->createAttributeSet($variation_type, 'test_color_attribute', [
+      'red' => 'Red',
+      'green' => 'Green',
+    ]);
+
+    // Reload the variation since we have new fields.
+    $this->variation = ProductVariation::load($this->variation->id());
+
+    // Get the product so we can append new variations.
+    $product = $this->variation->getProduct();
+
+    // Update first variation to have the attribute's value.
+    $this->variation->set('test_size_attribute', $size_attributes['6']->id());
+    $this->variation->set('test_color_attribute', $color_attributes['red']->id());
+    $this->variation->save();
+
+    $attribute_values_matrix = [
+      ['7', 'red'],
+      ['8', 'red'],
+      ['9', 'red'],
+      ['10', 'red'],
+      ['8', 'green'],
+      ['9', 'green'],
+      ['10', 'green'],
+    ];
+
+    $variations = [
+      $this->variation,
+    ];
+    // Generate variations off of the attributes values matrix.
+    foreach ($attribute_values_matrix as $key => $value) {
+      $variation = $this->createEntity('commerce_product_variation', [
+        'type' => $variation_type->id(),
+        'sku' => $this->randomMachineName(),
+        'price' => [
+          'amount' => 999,
+          'currency_code' => 'USD',
+        ],
+        'test_size_attribute' => $size_attributes[$value[0]]->id(),
+        'test_color_attribute' => $color_attributes[$value[1]]->id(),
+      ]);
+      $variations[] = $variation;
+      $product->variations->appendItem($variation);
+    }
+
+    $product->save();
+
+    $this->drupalGet($product->toUrl());
+    // Trigger AJAX by changing color attribute.
+    $this->drupalPostAjaxForm(NULL, [
+      'purchased_entity[0][attributes][test_size_attribute]' => $size_attributes['9']->id(),
+    ], 'purchased_entity[0][attributes][test_size_attribute]');
+    $this->assertAttributeExists('edit-purchased-entity-0-attributes-test-color-attribute', $color_attributes['red']->id());
+    $this->assertAttributeExists('edit-purchased-entity-0-attributes-test-color-attribute', $color_attributes['green']->id());
+
+    $this->drupalPostAjaxForm(NULL, [
+      'purchased_entity[0][attributes][test_size_attribute]' => $size_attributes['7']->id(),
+    ], 'purchased_entity[0][attributes][test_size_attribute]');
+    $this->assertAttributeExists('edit-purchased-entity-0-attributes-test-color-attribute', $color_attributes['red']->id());
+    $this->assertAttributeExists('edit-purchased-entity-0-attributes-test-color-attribute', $color_attributes['green']->id());
+
+    $this->drupalPostAjaxForm(NULL, [
+      'purchased_entity[0][attributes][test_color_attribute]' => $color_attributes['green']->id(),
+    ], 'purchased_entity[0][attributes][test_color_attribute]');
+    $this->assertAttributeDoesNotExist('edit-purchased-entity-0-attributes-test-size-attribute', $size_attributes['6']->id());
```

Comment:
@bojanz this fails. Should it? `6in` is technically not a valid attribute value when checking `Green` product variations. Or should it, because we're exploring all options.